### ISSUE TITLE
Backward policy implementation

### ISSUE THREAD

**Issue by agarkovv** [zmy1116/phylogfn#1]
Hello, and thank you for your amazing work on this project!

While exploring the implementation of phylogenetic trees, I had a question that I hope you can clarify:

Is it possible for the state representation to perform a backward transition using a uniform policy in the inner state $s_t$ and subsequently return to the same state $s_t$ via a forward policy?

From the implementation, it seems that a complete rooted tree is initially constructed, followed by backward sampling until the "all leaves" state is reached. I would greatly appreciate it if you could provide some insight into this process.

Thank you in advance for your time and explanation!

**Comment by zmy1116** [comment]
Hello, 

I don't quite understand your questions.. but maybe I can commment on some things you said..

> Is it possible for the state representation to perform a backward transition using a uniform policy in the inner state $$s_t$$  and subsequently return to the same state  $$s_t$$  via a forward policy?

If you are talking about tree generation specifically for phylogfn, we generate trees going `forward`. Backward policy is not used for trees generations.   As you can see the construciton MDP is a DAG. There is no trajectory goes forward then backward. 

If you are talking about in general can you design some other contruction process that goes back and forth, sure,  it's a free world. 

>  it seems that a complete rooted tree is initially constructed, followed by backward sampling until the "all leaves" state is reached. 

It has been a while and I have memory of a gold fish so I may not remember everything exactly. But in general:
- Trees are generated via forward policy. All possible generation paths are encasuplated within the MDP. And the forward policy is modeled with GFlowNet. 
- backward sampling policy is introduced for few reasons:
   -  To estimate sampling probability of a complete tree for a given forward policy.  We estimated it via importance sampling. (See appendix section B)
   -  To estimate MLL lower bound (See appendix section B)
   -  We use trajectory balance loss to train the gflownet.  It involves sampling p of a trajetory based on a backward sampling policy. (Refer to trajectory balance loss)

Thanks

**Comment by agarkovv** [comment]
Hello, and thank you for your response! I’d like to provide some additional details to clarify my question.

I’m not referring to tree generation specifically but rather the environment used for the GFlowNet. In a given state \( s_t \), we can transition to the next state by merging two rooted trees, resulting in a new state \( s_{t+1} \). My question focuses on the possible states after performing a backward action within this environment.

From my understanding of the paper and the implementation, the state \( s \) is represented as a forest of rooted trees. Does the backward action break any root of one of these rooted trees, or is it possible for this action to break a non-leaf vertex within a tree? 

So, if it is possible for a backward action to break any non-leaf vertex (e.g., some inner root of a subtree within a bigger tree), we cannot transition back to the starting state \( s_t \) by merging only roots. To recover the broken tree, we would need to merge a subtree back onto the broken branch.

I hope this clarification makes my question more precise. Apologies for not providing the full context initially, and thank you again for your time and explanation!

**Comment by zmy1116** [comment]
> is it possible for this action to break a non-leaf vertex within a tree (also not the current root)

No, if you break at an internal node, one rooted tree is splited into 3 trees instead of 2.  There is not any previous state $$s_{t-1}$$ that merge 3 trees to one. (I think...)

But also I think in this repo I only wrote the constructions of **unrooted trees**. That's why going backward the last state is different than the previous states 

https://github.com/zmy1116/phylogfn/blob/6a46f8682ff479961d66a367fec5ee9e798fc335/src/env/binary_tree_env_one_step_likelihood.py#L452-L462

So at last state, we DO ALLOW TO break at any edge of the full tree.  The way to think about it is that:
- Going **FORWARD**, we first generate a rooted tree, then remove the top root node 
- Going **BACKWARD**, given an unrooted tree, we first choose an edge, and add a root over the edge to fold the unrooted tree to a rooted tree, we then remove the root, resulting a forest of rooted trees, and going back...

There is an 1 to 1 correspondance between the set of forward policy generated trajectories and backward policy generated trajectories.

So when thinking about possible backward actions, you need to think about if there is any corresponding previous state/ forward action to arrive to the current state.

**Comment by agarkovv** [comment]
Hello, and thank you for your detailed explanations! I have a follow-up question regarding the last backward step in the environment.

From your response, you mentioned:

> At last state, we DO ALLOW to break at any edge of the full tree.

I am trying to reconcile this with the backward transition constraints. If we allow choosing any edge inside the tree, this could potentially break a non-leaf vertex, resulting in three separate trees instead of two. However, in a standard backward step, we are reversing a forward merge, which always combines exactly two trees.

* _How does choosing an arbitrary edge ensure consistency with the forward trajectory?_
If we pick an edge inside a subtree instead of at the last merge point, doesn’t this split the tree into three parts instead of two?
Wouldn’t that make it impossible to match a previous forward merge step?

* _Is there a mechanism ensuring that even if an arbitrary edge is chosen, the resulting state remains valid within the GFlowNet framework?_
How do we maintain the one-to-one correspondence between forward and backward trajectories in this case?

I’d appreciate any clarification on this aspect, especially if I’m missing something in the implementation. Thanks again for your time and help!

**Comment by zmy1116** [comment]
Hi,

Let's  agree that the goal is to generate **unrooted trees**.  
In a forward path:
- we start by constructing a rooted tree, by iteratively merging at the root of a pair of subtrees
- at the end, we **remove the root and flatten out** the two two edges into a single edge, can you visualize that?

So, for the corresponding backward trajectory
- The first step is to fold an unrooted tree to a rooted tree.  Can you visualize how this is done? We pick any of the edge, place a node over the edge, and fold it into a rooted tree.  This step add a new node in the tree and change the tree structure from an unrooted tree to a rooted tree. **No vertex is broken**
- then  we proceed by removing root node. 

Try to think about the process: 
- set of sequences -> merge to a rooted tree  -> bend the top to create an unrooted tree  -> unrooted tree
- set of sequences <-  iteratively choose a root node to break  <-  choose an edge to fold into a rooted tree <- unrooted tree

Let's first agree on this. if you have furhter question on consistency with gflownet and mathematical formulattion we can discuss more.

**Comment by agarkovv** [comment]
Thanks for the clear explanation! The key was the transition between rooted and unrooted trees—now I get it. Appreciate your help!
